TopMovies/Core/State/CategoriesPaginationState.swift

```diff
@@ -0,0 +1,84 @@
+  }
+}
+// MARK: - Reducer -
+extension CategoriesPaginationState {
+  static func reduce(action: Action, state: CategoriesPaginationState) -> CategoriesPaginationState {
+    var newPaginated = state.paginated
+    switch action {
+    // MARK: - Reduce MovieCategoriesAction
+    case let completedCategories as CompletedMovieCategoriesAction:
+      return CategoriesPaginationState(
+        paginated: completedCategories.relational
+          .mapValues { CategoryState.init(list: $0,
+                                          reload: CategoryState.defaultValue.reload,
+                                          loadMore: CategoryState.defaultValue.loadMore,
+                                          pageInfo: CategoryState.defaultValue.pageInfo) })
+    // MARK: - Reduce MoviesListActions
+    case let requestedMoviesList as RequestedMoviesListAction:
+      let categoryId = requestedMoviesList.categoryId
+      newPaginated
+        .updateValue(.init(list: (state.paginated[categoryId] ?? .defaultValue) .list,
+                           reload: requestedMoviesList.requestType == .reload ?
+                            .requested : .initial,
+                           loadMore: requestedMoviesList.requestType == .loadMore ?
+                            .requested : .initial,
+                           pageInfo: (state.paginated[categoryId] ?? .defaultValue) .pageInfo),
+                     forKey: categoryId)
+      return CategoriesPaginationState(paginated: newPaginated)
+    case let downloadingAction as DownloadingMoviesListAction:
+      let categoryId = downloadingAction.categoryId
+      newPaginated
+        .updateValue(.init(list: (state.paginated[categoryId] ?? .defaultValue) .list,
+                           reload: (state.paginated[categoryId] ?? .defaultValue)
+                            .reload.isRequested ? .downloading : .initial,
+                           loadMore: (state.paginated[categoryId] ?? .defaultValue)
+                            .loadMore.isRequested ? .downloading : .initial,
+                           pageInfo: (state.paginated[categoryId] ?? .defaultValue) .pageInfo),
+                     forKey: categoryId)
+      return CategoriesPaginationState(paginated: newPaginated)
+    case let completedMoviesList as CompletedMoviesListAction:
+      let categoryId = completedMoviesList.categoryId
+      var listIfLoadMore = (state.paginated[categoryId] ?? .defaultValue) .list
+      listIfLoadMore.append(contentsOf: completedMoviesList.list.map(\.id))
+      newPaginated
+        .updateValue(.init(list: (state.paginated[categoryId] ?? .defaultValue) .reload.isDownloading ?
+                            completedMoviesList.list.map(\.id) : listIfLoadMore,
+                           reload: .initial,
+                           loadMore: .initial,
+                           pageInfo: completedMoviesList.nextPage != nil ?
+                            .next(completedMoviesList.nextPage!) : .lastPage),
```

Comment:
Force unwrap is not really appreciated in any way, since now it does work correctly, but some minor mistake in some distant future will make the app crash (let's say you or someone else will rewrite it as `completedMoviesList.nextPage == nil`

I'd prefer to write it like that:
```swift
let pageInfo = completedMoviesList.nextPage.map(PageInfo.next) ?? PageInfo.lastPage
```